Expose moveFocus()

jaksenko opened the pull request:
This makes the library ready for menus that open on hover.

This is a great library that handles keyboard users and accessibility perfectly, but only when your submenu opens on click. In our company the menu must open on hover, which can be achieved simply by attaching mouse events to the element:
```
<li className={styles.menuItem} onMouseEnter={() => setIsOpen(true)} onMouseLeave={() => setIsOpen(false)}>
```
But once menu is open, hovering over items won't change focus element. This again can be done by attaching mouse event to submenu item:
```
onMouseEnter={(elem) => elem.target.focus()}
```
However, the internal state of this library won't change so once you switch back to keyboard and hit up/down arrow, an unexpected element is focused.

With this change, one can use `moveFocus` to change also internal state:
```
onMouseEnter={() => moveFocus(index)} // index from `.map(menuItem, index)`
```
so switching between keyboard and mouse is seamless.

<!-- Thanks for your interest in contributing! Please read the contributing guide (CONTRIBUTING.md) before submitting a PR. -->

WesCossick commented:
It looks like a few steps were missed from the [contributing guide](https://github.com/sparksuite/react-accessible-dropdown-menu-hook/blob/master/CONTRIBUTING.md).

jaksenko committed:
Expose moveFocus()

This makes the library ready for menus that open on hover.

This is a great library that handles keyboard users and accessibility perfectly, but only when your submenu opens on click. In our company the menu must open on hover, which can be achieved simply by attaching mouse events to the element:
```
<li className={styles.menuItem} onMouseEnter={() => setIsOpen(true)} onMouseLeave={() => setIsOpen(false)}>
```
But once menu is open, hovering over items won't change focus element. This again can be done by attaching mouse event to submenu item:
```
onMouseEnter={(elem) => elem.target.focus()}
```
However, the internal state of this library won't change so once you switch back to keyboard and hit up/down arrow, an unexpected element is focused.

With this change, one can use `moveFocus` to change also internal state:
```
onMouseEnter={() => moveFocus(index)} // index from `.map(menuItem, index)`
```
so switching between keyboard and mouse is seamless.

jaksenko commented:
> It looks like a few steps were missed from the [contributing guide](https://github.com/sparksuite/react-accessible-dropdown-menu-hook/blob/master/CONTRIBUTING.md).

Ah right! I hope I didn't miss anything now.

WesCossick commented on website/docs/getting-started/using.md:
````diff
@@ -5,7 +5,7 @@ title: Using
 
 ```jsx
-const { buttonProps, itemProps, isOpen, setIsOpen } = useDropdownMenu(numberOfItems);
+const { buttonProps, itemProps, isOpen, moveFocus, setIsOpen } = useDropdownMenu(numberOfItems);
````
So with the getting started example, right now we've documented every variable it returns, but as we add more, this probably isn't the best strategy. Instead, it's probably best if these basic examples only document the variables you'd need to use for simple use cases.

With that in mind, I've created #283 to remove `setIsOpen` from these deconstructions, and I'd want to leave this file unchanged in this PR.

```suggestion
const { buttonProps, itemProps, isOpen, setIsOpen } = useDropdownMenu(numberOfItems);
```

WesCossick commented on website/docs/getting-started/using.md:
```diff
@@ -20,6 +20,7 @@ Create the menu with the `role='menu'` property and spread `itemProps[x]` onto e
 <div className={isOpen ? 'visible' : ''} role='menu'>
     <a {...itemProps[0]} href='https://example.com'>Regular link</a>
     <a {...itemProps[1]} onClick={handleClick}>With click handler</a>
+    <a {...itemProps[2]} onClick={() => moveFocus(2)}>With mouse support</a>
```
See comment above about leaving this file unchanged.

```suggestion
```

WesCossick commented on website/docs/design/options.md:
````diff
@@ -13,7 +13,7 @@ Option | Explanation
 
 ```js
-const { buttonProps, itemProps, isOpen, setIsOpen } = useDropdownMenu(numberOfItems, {
+const { buttonProps, itemProps, isOpen, moveFocus, setIsOpen } = useDropdownMenu(numberOfItems, {
````
Also in #283, I plan to remove the variable instantiation entirely, since this page really only focuses on the options (passed as the second argument to `userDropdownMenu()`. In light of that, let's leave the file unchanged in this PR.

```suggestion
const { buttonProps, itemProps, isOpen, setIsOpen } = useDropdownMenu(numberOfItems, {
```

WesCossick commented on src/use-dropdown-menu.ts:
```diff
@@ -24,6 +24,7 @@ interface DropdownMenuResponse {
 		ref: React.RefObject<HTMLAnchorElement>;
 	}[];
 	readonly isOpen: boolean;
+	readonly moveFocus: (itemIndex: number) => void;
 	readonly setIsOpen: React.Dispatch<React.SetStateAction<boolean>>;
```
I would swap these two lines so that we keep `isOpen` and `setIsOpen` (which are logically related) together.

```suggestion
	readonly setIsOpen: React.Dispatch<React.SetStateAction<boolean>>;
	readonly moveFocus: (itemIndex: number) => void;
```

WesCossick commented on src/use-dropdown-menu.ts:
```diff
@@ -233,5 +234,5 @@ export default function useDropdownMenu(itemCount: number, options?: DropdownMen
 
 	// Return a listener for the button, individual list items, and the state of the menu
-	return { buttonProps, itemProps, isOpen, setIsOpen } as const;
+	return { buttonProps, itemProps, isOpen, moveFocus, setIsOpen } as const;
```
Similarly, I would swap `moveFocus` with `setIsOpen` so that the two "is open" related variables are grouped together.

```suggestion
	return { buttonProps, itemProps, isOpen, setIsOpen, moveFocus } as const;
```

WesCossick commented on README.md:
````diff
@@ -38,7 +38,7 @@ import useDropdownMenu from 'react-accessible-dropdown-menu-hook';
 
 ```tsx
-const { buttonProps, itemProps, isOpen, setIsOpen } = useDropdownMenu(numberOfItems);
+const { buttonProps, itemProps, isOpen, moveFocus, setIsOpen } = useDropdownMenu(numberOfItems);
````
See other comments about #283. In light of that issue, we can leave this file unchanged.

```suggestion
const { buttonProps, itemProps, isOpen, setIsOpen } = useDropdownMenu(numberOfItems);
```

WesCossick reviewed (changes requested):
I've left a handful of comments with suggested changes, most of them have to do with how we'll document the return value differently going forward (#283).

In addition to that, it looks like the "Lint code" job is failing. Here's how to address that: https://github.com/sparksuite/react-accessible-dropdown-menu-hook/blob/master/CONTRIBUTING.md#7-format-the-code.

Finally, it looks like the ["Return object" page](https://sparksuite.github.io/react-accessible-dropdown-menu-hook/docs/design/return-object) wasn't updated in the docs, but should be, since this PR introduces a new item in the return object.

jaksenko committed:
Apply suggestions from code review

Co-authored-by: Wes Cossick <WesCossick@users.noreply.github.com>

jaksenko committed:
Format the code properly

jaksenko committed:
Update website docs

jaksenko commented:
Thank you for your thorough review. Hopefully I didn't forget anything this time.

WesCossick committed:
Merge branch 'master' into patch-1

WesCossick reviewed (approved):
LGTM and all the CI checks passed. @corymharper do you have any thoughts on this before I merge it?

corymharper commented:
> LGTM and all the CI checks passed. @corymharper do you have any thoughts on this before I merge it?

The implementation is fairly straightforward and I don't see any eye glaring problems with it, probably that the tests  for this new function and the verbiage in the documentation could be polished, but that can be done in a separated PR.
